tests/security/ConcurrentAuthLimit/ConcurrentAuthLimit.cpp

```diff
@@ -0,0 +1,386 @@
+#include <dds/DCPS/RTPS/BaseMessageTypes.h>
+#include <dds/DCPS/RTPS/GuidGenerator.h>
+#include <dds/DCPS/RTPS/MessageTypes.h>
+#include <dds/DCPS/RTPS/ParameterListConverter.h>
+#include <dds/DCPS/RTPS/RtpsDiscovery.h>
+
+#include <dds/DCPS/security/framework/Properties.h>
+#include <dds/DCPS/security/AuthenticationBuiltInImpl.h>
+
+#include <dds/DCPS/transport/rtps_udp/RtpsUdpInst.h>
+#ifdef ACE_AS_STATIC_LIBS
+#include <dds/DCPS/transport/rtps_udp/RtpsUdp.h>
+#endif
+
+#include <dds/DCPS/LogAddr.h>
+
+#include <ace/SOCK_Dgram.h>
+#include <ace/SOCK_Dgram_Mcast.h>
+
+using namespace OpenDDS::DCPS;
+using namespace OpenDDS::RTPS;
+
+void append(DDS::PropertySeq& props, const char* name, const OpenDDS::DCPS::String& value, bool propagate = false)
+{
+  const DDS::Property_t prop = {name, value.c_str(), propagate};
+  const unsigned int len = props.length();
+  props.length(len + 1);
+  props[len] = prop;
+}
+
+bool receive(const OpenDDS::DCPS::GUID_t& participant_guid,
+             ACE_SOCK_Dgram_Mcast& multicast_socket)
+{
+  // Wait for an SPDP announcement before proceeding.
+  ACE_Message_Block buff(64 * 1024);
+
+  ACE_INET_Addr remote;
+  buff.reset();
+
+#ifdef ACE_LACKS_SENDMSG
+  const ssize_t bytes = multicast_socket.recv(buff.wr_ptr(), buff.space(), remote);
+#else
+  ACE_INET_Addr local;
+
+  iovec iov[1];
+  iov[0].iov_base = buff.wr_ptr();
+#ifdef _MSC_VER
+#pragma warning(push)
+  // iov_len is 32-bit on 64-bit VC++, but we don't want a cast here
+  // since on other platforms iov_len is 64-bit
+#pragma warning(disable : 4267)
+#endif
+  iov[0].iov_len = buff.space();
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+  const ssize_t bytes = multicast_socket.recv(iov, 1, remote, 0
+#if defined(ACE_RECVPKTINFO) || defined(ACE_RECVPKTINFO6)
+                                    , &local
+#endif
+                                    );
+#endif
+
+  buff.wr_ptr(bytes);
+
+  const Encoding encoding_plain_native(Encoding::KIND_XCDR1);
+  OpenDDS::DCPS::Serializer ser(&buff, encoding_plain_native);
+  Header header;
+  if (!(ser >> header)) {
+    return EXIT_FAILURE;
+  }
+
+  return make_part_guid(header.guidPrefix) == participant_guid;
+}
+
+void to_locator(const ACE_INET_Addr& addr, Locator_t& locator)
+{
+#ifdef ACE_HAS_IPV6
+  if (addr.get_type() == AF_INET6) {
+    locator.kind = LOCATOR_KIND_UDPv6;
+    struct sockaddr_in6* in6 = static_cast<struct sockaddr_in6*>(addr.get_addr());
+    ACE_OS::memcpy(reinterpret_cast<unsigned char*>(locator.address), &in6->sin6_addr, 16);
+  } else
+#endif
+  {
```

Comment:
Adding a `return;` to the end of the first block may be easier to read than this